cadc-dali-parquet/src/main/java/ca/nrc/cadc/dali/tables/parquet/ParquetReader.java

```diff
@@ -69,199 +69,248 @@
+import ca.nrc.cadc.dali.util.FloatArrayFormat;
+import ca.nrc.cadc.dali.util.FloatFormat;
+import ca.nrc.cadc.dali.util.Format;
+import ca.nrc.cadc.dali.util.FormatFactory;
+import ca.nrc.cadc.dali.util.IntArrayFormat;
+import ca.nrc.cadc.dali.util.IntegerFormat;
+import ca.nrc.cadc.dali.util.LongArrayFormat;
+import ca.nrc.cadc.dali.util.LongFormat;
+import ca.nrc.cadc.dali.util.StringFormat;
+import ca.nrc.cadc.dali.util.UTCTimestampFormat;
+import ca.nrc.cadc.dali.util.UUIDFormat;
+import ca.nrc.cadc.io.ByteCountInputStream;
+
+import java.io.File;
+import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.ByteBuffer;
-import java.util.HashMap;
-import java.util.Map;
-import org.apache.avro.generic.GenericRecord;
+import java.io.OutputStream;
+import java.io.RandomAccessFile;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 import org.apache.log4j.Logger;
-import org.apache.parquet.avro.AvroParquetReader;
 import org.apache.parquet.hadoop.ParquetFileReader;
 import org.apache.parquet.hadoop.metadata.ParquetMetadata;
 import org.apache.parquet.io.InputFile;
-import org.apache.parquet.io.SeekableInputStream;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
 import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.PrimitiveType;
 import org.apache.parquet.schema.Type;
 
-
-//TODO: Temporary implementation
 public class ParquetReader {
 
     private static final Logger log = Logger.getLogger(ParquetReader.class);
 
-    public TableShape read(InputStream inputStream) throws IOException {
+    private static final long BYTE_LIMIT = 1024 * 1024L; // 1 MiB
 
-        InputFile inputFile = getInputFile(inputStream);
-        int recordCount = 0;
-        int columnCount = 0;
+    private static final FormatFactory formatFactory = new FormatFactory();
 
-        ParquetMetadata metadata = ParquetFileReader.open(inputFile).getFooter();
-        MessageType schema = metadata.getFileMetaData().getSchema();
-        columnCount = schema.getFieldCount();
+    public VOTableDocument read(RandomAccessFile file) throws IOException {
+        ParquetInputFile parquetInputFile = new ParquetInputFile(file);
+        return read(parquetInputFile);
+    }
 
-        String votable = metadata.getFileMetaData().getKeyValueMetaData().get("IVOA.VOTable-Parquet.content");
-        log.debug("VOTable: " + votable);
+    public VOTableDocument read(RandomAccessSource randomAccessSource) throws IOException {
+        RandomAccessSeekableInputStream parquetInputFile = new RandomAccessSeekableInputStream(randomAccessSource);
+        return read(parquetInputFile);
+    }
 
-        VOTableReader voTableReader = new VOTableReader();
-        VOTableDocument voTableDocument = voTableReader.read(votable);
+    public VOTableDocument read(InputStream inputStream) throws IOException {
```

Comment:
The `read` methods here should have javadocs and this one must indicate that it reads the stream fully and stores it in a local temporary file before returning.

I do not think the byte limit and ByteCountInputStream should be added here. The actual limit will depend on context so the caller can/should wrap the input stream if necessar. For example, the BasicUploadManager in cadc-tap-server already adds a ByteCountInputStream limitation.